### PR TITLE
fix: Cannot read properties of null reading "write"

### LINKED ISSUE #56: cannot read property 'write' of null
im using m1 mac

this is my init.vim right now
Screen Shot 2021-09-02 at 10 04 57 PM

and the error is
```[coc.nvim] Error on activate extension coc-discord-rpc: Cannot read property 'write' of null```

### LINKED ISSUE #59: Cannot read properties of null reading "write"
Been coding on nvim and I installed this Plugin until encountered this error

Screenshot_20211119_165633

### PR DESCRIPTION
**Pull request by xhayper**
Cause:
   This was caused by 2 issue stacking over each other
   1. neovim OR coc set the `$TMPDIR` to `/var/folders/tg/**/T/nvim**`
   2. macOS returning `/var/folders/tg/**/T/**` via `$TMPDIR` when the real tmp directory is `/private/var/folders/tg/**/T/**`
 
Fix:
  What we do is we loop through the path variable that `discord-rpc` search and fix the path,
     1. Loop through the path variable
     2. Turn the path into real path
     3. check if it contains nvim
     4. if yes, get the parent directory
     5. if no, continue
     6. set that variable to modified one

Note:
   - This was tested on a macOS machine, running on macOS 12.5, This still need to be tested on linux machine too.
   - on macOS, `/var` is a symlink to `/private/var`, i am not sure if `net` package can handle symlink properly, so the i did the `fs.realPath`, I am not taking any risk either.

Fixes #59 
Fixes #56

### CODE REVIEW

**Comment by xhayper** [comment]
Screen Shot 2022-07-24 at 6 23 33 PM
Screen Shot 2022-07-24 at 6 27 22 PM

**Comment by leonardssh** [comment]
Thanks for this pR, I'm glad you found a solution to this problem, I'm going to make a Linux VM environment so I can test it too. 

I'll keep this PR open until we see if it's right.

File: .gitignore [hunk 1]
```diff
@@ -109,3 +109,6 @@ lib
 old_src
 
 build.sh
+
+# lock file
+yarn.lock
```
Comment (leonardssh) [inline .gitignore:114]: Actually, we need yarn.lock file to be present to be able to cache him for github workflow.

**Review by leonardssh — CHANGES_REQUESTED** [review]

File: .gitignore (hunk 1 above)
Comment (xhayper) [reply .gitignore:114]: I see, i will get the yarn.lock back.

**Review by leonardssh — APPROVED** [review]

**Comment by xhayper** [comment]
I gonna test this on an Arch Linux VM with KDE Plasma and SDDM.

**Comment by xhayper** [comment]
Works great.
image

**Comment by xhayper** [comment]
Ah... `discord-ipc-{number}` in this case is inside `XDG_RUNTIME_DIR`, which we have no problem with, BUT HOWEVER
image
`TMPDIR` is wrong here, Which might explain why some people got the error, and some don't

**Comment by leonardssh** [comment]
> Ah... `discord-ipc-{number}` in this case is inside `XDG_RUNTIME_DIR`, which we have no problem with, BUT HOWEVER image `TMPDIR` is wrong here, Which might explain why people who are using `X` doesn't have the error.

It makes perfect sense.

**Comment by xhayper** [comment]
Trying this on Wayland version of plasma.

**Comment by xhayper** [comment]
My assumption are correct.
image

**Comment by xhayper** [comment]
Yup! This patch fix it!
image

**Comment by xhayper** [comment]
Summary:
   discord-rpc will breaks under Wayland with coc / neovim.
   Why? coc / neovim set $TMPDIR to `/tmp/nvim**`

**Comment by xhayper** [comment]
Alright, I got tired of `discordjs/RPC`, I made my own fork of `discordjs/RPC` that's up to date, use typescript and can specify your own way to find the ipc path.

**Comment by xhayper** [comment]
If you want, I can port to my discord-rpc for this project in separate PR.

**Comment by leonardssh** [comment]
A PR in this regard would be welcome. I also want to port my [vscode extension](https://github.com/leonardssh/vscord) to your discord-rpc fork.

**Comment by xhayper** [comment]
Alright! I just need someone to test if the extension works under Snap version of discord.

**Comment by xhayper** [comment]
Yep, the thing works with snap versions of discord.

**Comment by KarlWithK** [comment]
Thank you for your work @xhayper!

**Pull request closed without merging**